Examples/Framework/ML/src/NeuralCalibrator.cpp

```diff
@@ -103,80 +103,71 @@ void ActsExamples::NeuralCalibrator::calibrate(
+
+  const Acts::BoundVector& fpar = measurement.fullParameters();
+  const Acts::BoundMatrix& fcov = measurement.fullCovariance();
+
+  Acts::Vector3 dir = Acts::makeDirectionFromPhiTheta(fpar[Acts::eBoundPhi],
+                                                      fpar[Acts::eBoundTheta]);
+  Acts::Vector3 globalPosition = referenceSurface.localToGlobal(
+      gctx, fpar.segment<2>(Acts::eBoundLoc0), dir);
+
+  // Rotation matrix. When applied to global coordinates, they
+  // are rotated into the local reference frame of the
+  // surface. Note that this such a rotation can be found by
+  // inverting a matrix whose columns correspond to the
+  // coordinate axes of the local coordinate system.
+  Acts::RotationMatrix3 rot =
+      referenceSurface.referenceFrame(gctx, globalPosition, dir).inverse();
+  std::pair<double, double> angles =
+      Acts::VectorHelpers::incidentAngles(dir, rot);
+
+  input[iInput++] = angles.first;
+  input[iInput++] = angles.second;
+  input[iInput++] = fpar[Acts::eBoundLoc0];
+  input[iInput++] = fpar[Acts::eBoundLoc1];
+  input[iInput++] = fcov(Acts::eBoundLoc0, Acts::eBoundLoc0);
+  input[iInput++] = fcov(Acts::eBoundLoc1, Acts::eBoundLoc1);
+  if (iInput != m_nInputs) {
+    throw std::runtime_error("Expected input size of " +
+                             std::to_string(m_nInputs) +
+                             ", got: " + std::to_string(iInput));
+  }
+
+  // Input is a single row, hence .front()
+  std::vector<float> output = m_model.runONNXInference(inputBatch).front();
+  // Assuming 2-D measurements, the expected params structure is:
+  // [           0,    nComponent[ --> priors
+  // [  nComponent,  3*nComponent[ --> means
+  // [3*nComponent,  5*nComponent[ --> variances
+  std::size_t nParams = 5 * m_nComponents;
+  if (output.size() != nParams) {
+    throw std::runtime_error("Got output vector of size " +
+                             std::to_string(output.size()) +
+                             ", expected size " + std::to_string(nParams));
+  }
+
+  // Most probable value computation of mixture density
+  std::size_t iMax = 0;
+  if (m_nComponents > 1) {
+    iMax = std::distance(
+        output.begin(),
+        std::max_element(output.begin(), output.begin() + m_nComponents));
+  }
+  std::size_t iLoc0 = m_nComponents + iMax * 2;
+  std::size_t iVar0 = 3 * m_nComponents + iMax * 2;
+
+  Measurement measurementCopy = measurement;
+  measurementCopy.parameters()[Acts::eBoundLoc0] = output[iLoc0];
+  measurementCopy.parameters()[Acts::eBoundLoc1] = output[iLoc0 + 1];
+  measurementCopy.covariance()(Acts::eBoundLoc0, Acts::eBoundLoc0) =
+      output[iVar0];
+  measurementCopy.covariance()(Acts::eBoundLoc1, Acts::eBoundLoc1) =
+      output[iVar0 + 1];
+
+  trackState.allocateCalibrated(measurementCopy.size());
+  trackState.effectiveCalibrated() = measurementCopy.parameters();
+  trackState.effectiveCalibratedCovariance() = measurementCopy.covariance();
```

Comment:
If possible, I'd like to avoid increasing usage of the dynamic types here. I'm not sure how Eigen implements assignment between these, but this could be done with compile-time sizes in any case.

Reply:
Switched to access directly by measurement component index and the assignment to template unrolling.